File: ql/time/period.cpp
/* -*- mode: c++; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*- */

/*
 Copyright (C) 2004, 2005, 2006, 2007, 2008, 2014 Ferdinando Ametrano
 Copyright (C) 2006 Katiuscia Manzoni
 Copyright (C) 2000, 2001, 2002, 2003 RiskMap srl
 Copyright (C) 2003, 2004, 2005, 2006, 2008 StatPro Italia srl
 Copyright (C) 2014 Paolo Mazzocchi

 This file is part of QuantLib, a free-software/open-source library
 for financial quantitative analysts and developers - http://quantlib.org/

 QuantLib is free software: you can redistribute it and/or modify it
 under the terms of the QuantLib license.  You should have received a
 copy of the license along with this program; if not, please email
 <quantlib-dev@lists.sf.net>. The license is also available online at
 <http://quantlib.org/license.shtml>.

 This program is distributed in the hope that it will be useful, but WITHOUT
 ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 FOR A PARTICULAR PURPOSE.  See the license for more details.
*/

#include <ql/time/period.hpp>
#include <ql/errors.hpp>

namespace QuantLib {

    Period::Period(Frequency f) {
        switch (f) {
          case NoFrequency:
            // same as Period()
            units_ = Days;
            length_ = 0;
            break;
          case Once:
            units_ = Years;
            length_ = 0;
            break;
          case Annual:
            units_ = Years;
            length_ = 1;
            break;
          case Semiannual:
          case EveryFourthMonth:
          case Quarterly:
          case Bimonthly:
          case Monthly:
            units_ = Months;
            length_ = 12/f;
            break;
          case EveryFourthWeek:
          case Biweekly:
          case Weekly:
            units_ = Weeks;
            length_ = 52/f;
            break;
          case Daily:
            units_ = Days;
            length_ = 1;
            break;
          case OtherFrequency:
            QL_FAIL("unknown frequency");  // no point in showing 999...
          default:
            QL_FAIL("unknown frequency (" << Integer(f) << ")");
        }
    }

    Frequency Period::frequency() const {
        // unsigned version
        Size length = std::abs(length_);

        if (length==0) {
            if (units_==Years) return Once;
            return NoFrequency;
        }

        switch (units_) {
          case Years:
            if (length == 1)
                return Annual;
            else
                return OtherFrequency;
          case Months:
            if (12%length == 0 && length <= 12)
                return Frequency(12/length);
            else
                return OtherFrequency;
          case Weeks:
            if (length==1)
                return Weekly;
            else if (length==2)
                return Biweekly;
            else if (length==4)
                return EveryFourthWeek;
            else
                return OtherFrequency;
          case Days:
            if (length==1)
                return Daily;
            else
                return OtherFrequency;
          default:
            QL_FAIL("unknown time unit (" << Integer(units_) << ")");
        }
    }

    void Period::normalize() {
        if (length_ == 0) {
            units_ = Days;
        } else {
            switch (units_) {
              case Months:
                if ((length_ % 12) == 0) {
                    length_ /= 12;
                    units_ = Years;
                }
                break;
              case Days:
              case Weeks:
              case Years:
                break;
              default:
                QL_FAIL("unknown time unit (" << Integer(units_) << ")");
            }
        }
    }

    Period& Period::operator+=(const Period& p) {

        if (length_==0) {
            length_ = p.length();
            units_ = p.units();
        } else if (units_==p.units()) {
            // no conversion needed
            length_ += p.length();
        } else {
            switch (units_) {

              case Years:
                switch (p.units()) {
                  case Months:
                    units_ = Months;
                    length_ = length_*12 + p.length();
                    break;
                  case Weeks:
                  case Days:
                    QL_REQUIRE(p.length()==0,
                               "impossible addition between " << *this <<
                               " and " << p);
                    break;
                  default:
                    QL_FAIL("unknown time unit (" << Integer(p.units()) << ")");
                }
                break;

              case Months:
                switch (p.units()) {
                  case Years:
                    length_ += p.length()*12;
                    break;
                  case Weeks:
                  case Days:
                    QL_REQUIRE(p.length()==0,
                               "impossible addition between " << *this <<
                               " and " << p);
                    break;
                  default:
                    QL_FAIL("unknown time unit (" << Integer(p.units()) << ")");
                }
                break;

              case Weeks:
                switch (p.units()) {
                  case Days:
                    units_ = Days;
                    length_ = length_*7 + p.length();
                    break;
                  case Years:
                  case Months:
                    QL_REQUIRE(p.length()==0,
                               "impossible addition between " << *this <<
                               " and " << p);
                    break;
                  default:
                    QL_FAIL("unknown time unit (" << Integer(p.units()) << ")");
                }
                break;

              case Days:
                switch (p.units()) {
                  case Weeks:
                    length_ += p.length()*7;
                    break;
                  case Years:
                  case Months:
                    QL_REQUIRE(p.length()==0,
                               "impossible addition between " << *this <<
                               " and " << p);
                    break;
                  default:
                    QL_FAIL("unknown time unit (" << Integer(p.units()) << ")");
                }
                break;

              default:
                QL_FAIL("unknown time unit (" << Integer(units_) << ")");
            }
        }

        return *this;
    }

    Period& Period::operator-=(const Period& p) {
        return operator+=(-p);
    }

    Period& Period::operator*=(Integer n) {
        length_ *= n;
        return *this;
    }

    Period& Period::operator/=(Integer n) {
        QL_REQUIRE(n != 0, "cannot be divided by zero");
        if (length_ % n == 0) {
            // keep the original units. If the user created a
            // 24-months period, he'll probably want a 12-months one
            // when he halves it.
            length_ /= n;
        } else {
            // try
            TimeUnit units = units_;
            Integer length = length_;
            switch (units) {
              case Years:
                length *= 12;
                units = Months;
                break;
              case Weeks:
                length *= 7;
                units = Days;
                break;
              default:
                ;
            }
            QL_REQUIRE(length % n == 0,
                       *this << " cannot be divided by " << n);
            length_ = length/n;
            units_ = units;
        }
        return *this;
    }


    namespace {

        std::pair<Integer,Integer> daysMinMax(const Period& p) {
            switch (p.units()) {
              case Days:
                return std::make_pair(p.length(), p.length());
              case Weeks:
                return std::make_pair(7*p.length(), 7*p.length());
              case Months:
                return std::make_pair(28*p.length(), 31*p.length());
              case Years:
                return std::make_pair(365*p.length(), 366*p.length());
              default:
                QL_FAIL("unknown time unit (" << Integer(p.units()) << ")");
            }
        }

    }

    Real years(const Period& p) {
        if (p.length()==0) return 0.0;

        switch (p.units()) {
          case Days:
            QL_FAIL("cannot convert Days into Years");
          case Weeks:
            QL_FAIL("cannot convert Weeks into Years");
          case Months:
              return p.length()/12.0;
          case Years:
              return p.length();
          default:
            QL_FAIL("unknown time unit (" << Integer(p.units()) << ")");
        }
    }

    Real months(const Period& p) {
        if (p.length()==0) return 0.0;

        switch (p.units()) {
          case Days:
            QL_FAIL("cannot convert Days into Months");
          case Weeks:
            QL_FAIL("cannot convert Weeks into Months");
          case Months:
              return p.length();
          case Years:
              return p.length()*12.0;
          default:
            QL_FAIL("unknown time unit (" << Integer(p.units()) << ")");
        }
    }

    Real weeks(const Period& p) {
        if (p.length()==0) return 0.0;

        switch (p.units()) {
          case Days:
              return p.length()/7.0;
          case Weeks:
              return p.length();
          case Months:
            QL_FAIL("cannot convert Months into Weeks");
          case Years:
            QL_FAIL("cannot convert Years into Weeks");
          default:
            QL_FAIL("unknown time unit (" << Integer(p.units()) << ")");
        }
    }

    Real days(const Period& p) {
        if (p.length()==0) return 0.0;

        switch (p.units()) {
          case Days:
              return p.length();
          case Weeks:
              return p.length()*7.0;
          case Months:
            QL_FAIL("cannot convert Months into Days");
          case Years:
            QL_FAIL("cannot convert Years into Days");
          default:
            QL_FAIL("unknown time unit (" << Integer(p.units()) << ")");
        }
    }

    bool operator<(const Period& p1, const Period& p2) {

        // special cases
        if (p1.length() == 0)
            return p2.length() > 0;
        if (p2.length() == 0)
            return p1.length() < 0;

        // exact comparisons
        if (p1.units() == p2.units())
            return p1.length() < p2.length();
        if (p1.units() == Months && p2.units() == Years)
            return p1.length() < 12*p2.length();
        if (p1.units() == Years && p2.units() == Months)
            return 12*p1.length() < p2.length();

        // inexact comparisons (handled by converting to days and using limits)
        std::pair<Integer, Integer> p1lim = daysMinMax(p1);
        std::pair<Integer, Integer> p2lim = daysMinMax(p2);

        if (p1lim.second < p2lim.first)
            return true;
        else if (p1lim.first > p2lim.second)
            return false;
        else
            QL_FAIL("undecidable comparison between " << p1 << " and " << p2);
    }


    Period operator+(const Period& p1, const Period& p2) {
        Period result = p1;
        result += p2;
        return result;
    }

    Period operator-(const Period& p1, const Period& p2) {
        return p1+(-p2);
    }

    Period operator/(const Period& p, Integer n) {
        Period result = p;
        result /= n;
        return result;
    }

    // period formatting

    std::ostream& operator<<(std::ostream& out, const Period& p) {
        return out << io::short_period(p);
    }

    namespace detail {

        std::ostream& operator<<(std::ostream& out,
                                 const long_period_holder& holder) {
            Integer n = holder.p.length();
            switch (holder.p.units()) {
              case Days:
                return out << n << (n == 1 ? " day" : " days");
              case Weeks:
                return out << n << (n == 1 ? " week" : " weeks");
              case Months:
                return out << n << (n == 1 ? " month" : " months");
              case Years:
                return out << n << (n == 1 ? " year" : " years");
              default:
                QL_FAIL("unknown time unit (" << Integer(holder.p.units()) << ")");
            }
        }

        std::ostream& operator<<(std::ostream& out,
                                 const short_period_holder& holder) {
            Integer n = holder.p.length();
            switch (holder.p.units()) {
              case Days:
                return out << n << "D";
              case Weeks:
                return out << n << "W";
              case Months:
                return out << n << "M";
              case Years:
                return out << n << "Y";
              default:
                QL_FAIL("unknown time unit (" << Integer(holder.p.units()) << ")");
            }
        }

    }

    namespace io {

        detail::long_period_holder long_period(const Period& p) {
            return detail::long_period_holder(p);
        }

        detail::short_period_holder short_period(const Period& p) {
            return detail::short_period_holder(p);
        }

    }

}

File: quantlib/time/period.py
from __future__ import annotations

from quantlib.time.frequency import Frequency
from quantlib.time.timeunit import TimeUnit


class Period:
    def __init__(self, n: int | Frequency = 0, units: TimeUnit = TimeUnit.Days):
        if isinstance(n, Frequency):
            self._init_from_frequency(n)
        else:
            self._length = n
            self._units = units

    def _init_from_frequency(self, f: Frequency) -> None:
        if f == Frequency.NoFrequency:
            self._units = TimeUnit.Days
            self._length = 0
        elif f == Frequency.Once:
            self._units = TimeUnit.Years
            self._length = 0
        elif f == Frequency.Annual:
            self._units = TimeUnit.Years
            self._length = 1
        elif f in (
            Frequency.Semiannual,
            Frequency.EveryFourthMonth,
            Frequency.Quarterly,
            Frequency.Bimonthly,
            Frequency.Monthly,
        ):
            self._units = TimeUnit.Months
            self._length = 12 // int(f)
        elif f in (Frequency.EveryFourthWeek, Frequency.Biweekly, Frequency.Weekly):
            self._units = TimeUnit.Weeks
            self._length = 52 // int(f)
        elif f == Frequency.Daily:
            self._units = TimeUnit.Days
            self._length = 1
        elif f == Frequency.OtherFrequency:
            raise ValueError("unknown frequency")
        else:
            raise ValueError(f"unknown frequency ({int(f)})")

    def length(self) -> int:
        return self._length

    def units(self) -> TimeUnit:
        return self._units

    def frequency(self) -> Frequency:
        length = abs(self._length)

        if length == 0:
            if self._units == TimeUnit.Years:
                return Frequency.Once
            return Frequency.NoFrequency

        if self._units == TimeUnit.Years:
            return Frequency.Annual if length == 1 else Frequency.OtherFrequency
        elif self._units == TimeUnit.Months:
            if 12 % length == 0 and length <= 12:
                return Frequency(12 // length)
            return Frequency.OtherFrequency
        elif self._units == TimeUnit.Weeks:
            if length == 1:
                return Frequency.Weekly
            elif length == 2:
                return Frequency.Biweekly
            elif length == 4:
                return Frequency.EveryFourthWeek
            return Frequency.OtherFrequency
        elif self._units == TimeUnit.Days:
            return Frequency.Daily if length == 1 else Frequency.OtherFrequency
        else:
            raise ValueError(f"unknown time unit ({int(self._units)})")

    def normalize(self) -> None:
        if self._length == 0:
            self._units = TimeUnit.Days
        elif self._units == TimeUnit.Months:
            if self._length % 12 == 0:
                self._length //= 12
                self._units = TimeUnit.Years
        elif self._units == TimeUnit.Days:
            if self._length % 7 == 0:
                self._length //= 7
                self._units = TimeUnit.Weeks

    def normalized(self) -> Period:
        p = Period(self._length, self._units)
        p.normalize()
        return p

    def __iadd__(self, other: Period) -> Period:
        if self._length == 0:
            self._length = other._length
            self._units = other._units
        elif self._units == other._units:
            self._length += other._length
        else:
            self._iadd_different_units(other)
        return self

    def _iadd_different_units(self, other: Period) -> None:
        if self._units == TimeUnit.Years:
            if other._units == TimeUnit.Months:
                self._units = TimeUnit.Months
                self._length = self._length * 12 + other._length
            elif other._units in (TimeUnit.Weeks, TimeUnit.Days):
                if other._length != 0:
                    raise ValueError(f"impossible addition between {self} and {other}")
            else:
                raise ValueError(f"unknown time unit ({int(other._units)})")
        elif self._units == TimeUnit.Months:
            if other._units == TimeUnit.Years:
                self._length += other._length * 12
            elif other._units in (TimeUnit.Weeks, TimeUnit.Days):
                if other._length != 0:
                    raise ValueError(f"impossible addition between {self} and {other}")
            else:
                raise ValueError(f"unknown time unit ({int(other._units)})")
        elif self._units == TimeUnit.Weeks:
            if other._units == TimeUnit.Days:
                self._units = TimeUnit.Days
                self._length = self._length * 7 + other._length
            elif other._units in (TimeUnit.Years, TimeUnit.Months):
                if other._length != 0:
                    raise ValueError(f"impossible addition between {self} and {other}")
            else:
                raise ValueError(f"unknown time unit ({int(other._units)})")
        elif self._units == TimeUnit.Days:
            if other._units == TimeUnit.Weeks:
                self._length += other._length * 7
            elif other._units in (TimeUnit.Years, TimeUnit.Months):
                if other._length != 0:
                    raise ValueError(f"impossible addition between {self} and {other}")
            else:
                raise ValueError(f"unknown time unit ({int(other._units)})")
        else:
            raise ValueError(f"unknown time unit ({int(self._units)})")

    def __isub__(self, other: Period) -> Period:
        self += -other
        return self

    def __imul__(self, n: int) -> Period:
        self._length *= n
        return self

    def __itruediv__(self, n: int) -> Period:
        if n == 0:
            raise ValueError("cannot be divided by zero")
        if self._length % n == 0:
            self._length //= n
        else:
            units = self._units
            length = self._length
            if units == TimeUnit.Years:
                length *= 12
                units = TimeUnit.Months
            elif units == TimeUnit.Weeks:
                length *= 7
                units = TimeUnit.Days
            if length % n != 0:
                raise ValueError(f"{self} cannot be divided by {n}")
            self._length = length // n
            self._units = units
        return self

    def __neg__(self) -> Period:
        return Period(-self._length, self._units)

    def __add__(self, other: Period) -> Period:
        result = Period(self._length, self._units)
        result += other
        return result

    def __sub__(self, other: Period) -> Period:
        return self + (-other)

    def __mul__(self, n: int) -> Period:
        return Period(n * self._length, self._units)

    def __rmul__(self, n: int) -> Period:
        return Period(n * self._length, self._units)

    def __truediv__(self, n: int) -> Period:
        result = Period(self._length, self._units)
        result /= n
        return result

    def __eq__(self, other: object) -> bool:
        if not isinstance(other, Period):
            return NotImplemented
        return not (self < other or other < self)

    def __ne__(self, other: object) -> bool:
        if not isinstance(other, Period):
            return NotImplemented
        return not (self == other)

    def __lt__(self, other: Period) -> bool:
        # special cases
        if self._length == 0:
            return other._length > 0
        if other._length == 0:
            return self._length < 0

        # exact comparisons
        if self._units == other._units:
            return self._length < other._length
        if self._units == TimeUnit.Months and other._units == TimeUnit.Years:
            return self._length < 12 * other._length
        if self._units == TimeUnit.Years and other._units == TimeUnit.Months:
            return 12 * self._length < other._length
        if self._units == TimeUnit.Days and other._units == TimeUnit.Weeks:
            return self._length < 7 * other._length
        if self._units == TimeUnit.Weeks and other._units == TimeUnit.Days:
            return 7 * self._length < other._length

        # inexact comparisons
        p1_min, p1_max = _days_min_max(self)
        p2_min, p2_max = _days_min_max(other)

        if p1_max < p2_min:
            return True
        elif p1_min > p2_max:
            return False
        else:
            raise ValueError(f"undecidable comparison between {self} and {other}")

    def __le__(self, other: Period) -> bool:
        return not (other < self)

    def __gt__(self, other: Period) -> bool:
        return other < self

    def __ge__(self, other: Period) -> bool:
        return not (self < other)

    def __str__(self) -> str:
        _unit_short = {
            TimeUnit.Days: "D",
            TimeUnit.Weeks: "W",
            TimeUnit.Months: "M",
            TimeUnit.Years: "Y",
            TimeUnit.Hours: "h",
            TimeUnit.Minutes: "m",
            TimeUnit.Seconds: "s",
        }
        return f"{self._length}{_unit_short.get(self._units, '?')}"

    def __repr__(self) -> str:
        return f"Period({self._length}, {self._units!r})"


def _days_min_max(p: Period) -> tuple[int, int]:
    length = p.length()
    units = p.units()
    if units == TimeUnit.Days:
        return (length, length)
    elif units == TimeUnit.Weeks:
        return (7 * length, 7 * length)
    elif units == TimeUnit.Months:
        return (28 * length, 31 * length)
    elif units == TimeUnit.Years:
        return (365 * length, 366 * length)
    else:
        raise ValueError(f"unknown time unit ({int(units)})")


def years(p: Period) -> float:
    if p.length() == 0:
        return 0.0
    if p.units() == TimeUnit.Days:
        raise ValueError("cannot convert Days into Years")
    if p.units() == TimeUnit.Weeks:
        raise ValueError("cannot convert Weeks into Years")
    if p.units() == TimeUnit.Months:
        return p.length() / 12.0
    if p.units() == TimeUnit.Years:
        return float(p.length())
    raise ValueError(f"unknown time unit ({int(p.units())})")


def months(p: Period) -> float:
    if p.length() == 0:
        return 0.0
    if p.units() == TimeUnit.Days:
        raise ValueError("cannot convert Days into Months")
    if p.units() == TimeUnit.Weeks:
        raise ValueError("cannot convert Weeks into Months")
    if p.units() == TimeUnit.Months:
        return float(p.length())
    if p.units() == TimeUnit.Years:
        return p.length() * 12.0
    raise ValueError(f"unknown time unit ({int(p.units())})")


def weeks(p: Period) -> float:
    if p.length() == 0:
        return 0.0
    if p.units() == TimeUnit.Days:
        return p.length() / 7.0
    if p.units() == TimeUnit.Weeks:
        return float(p.length())
    if p.units() == TimeUnit.Months:
        raise ValueError("cannot convert Months into Weeks")
    if p.units() == TimeUnit.Years:
        raise ValueError("cannot convert Years into Weeks")
    raise ValueError(f"unknown time unit ({int(p.units())})")


def days(p: Period) -> float:
    if p.length() == 0:
        return 0.0
    if p.units() == TimeUnit.Days:
        return float(p.length())
    if p.units() == TimeUnit.Weeks:
        return p.length() * 7.0
    if p.units() == TimeUnit.Months:
        raise ValueError("cannot convert Months into Days")
    if p.units() == TimeUnit.Years:
        raise ValueError("cannot convert Years into Days")
    raise ValueError(f"unknown time unit ({int(p.units())})")
